staging/src/github.com/openshift/oc/pkg/cli/version/version.go

```diff
@@ -1,25 +1,185 @@
+
+	genericclioptions.IOStreams
+}
+
+func NewVersionOptions(ioStreams genericclioptions.IOStreams) *VersionOptions {
+	return &VersionOptions{
+		IOStreams: ioStreams,
+	}
+
+}
+
+func NewCmdVersion(fullName string, f cmdutil.Factory, ioStreams genericclioptions.IOStreams) *cobra.Command {
+	o := NewVersionOptions(ioStreams)
 	cmd := &cobra.Command{
-		Use:   "version",
-		Short: "Display version",
-		Long:  "Display version",
+		Use:     "version",
+		Short:   "Print the client and server version numbers",
+		Long:    "Print the client and server version numbers for the current context",
+		Example: fmt.Sprintf(versionExample, fullName),
 		Run: func(cmd *cobra.Command, args []string) {
-			fmt.Fprintf(out, "%s %v\n", fullName, versionInfo)
+			cmdutil.CheckErr(o.Complete(f, cmd))
+			cmdutil.CheckErr(o.Validate())
+			cmdutil.CheckErr(o.Run())
 		},
 	}
-
+	cmd.Flags().BoolVar(&o.ClientOnly, "client", o.ClientOnly, "Client version only (no server required).")
+	cmd.Flags().BoolVar(&o.ServerOnly, "server", o.ServerOnly, "Servers version only.")
+	cmd.Flags().BoolVar(&o.All, "all", o.All, "Print full version information.")
+	cmd.Flags().StringVarP(&o.Output, "output", "o", o.Output, "One of 'yaml' or 'json'.")
 	return cmd
 }
+
+func (o *VersionOptions) Complete(f cmdutil.Factory, cmd *cobra.Command) error {
+	var err error
+	o.discoveryClient, err = f.ToDiscoveryClient()
+	// if we had an empty rest.Config, continue and just print out client information.
+	// if we had an error other than being unable to build a rest.Config, fail.
+	if err != nil && !clientcmd.IsEmptyConfig(err) {
+		return err
+	}
+	clientConfig, err := f.ToRESTConfig()
+	if err != nil && !clientcmd.IsEmptyConfig(err) {
+		return err
+	}
+	o.oClient, err = configv1client.NewForConfig(clientConfig)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (o *VersionOptions) Validate() error {
+	if o.Output != "" && o.Output != "yaml" && o.Output != "json" {
+		return errors.New(`--output must be 'yaml' or 'json'`)
+	}
+
+	return nil
+}
+
+func (o *VersionOptions) Run() error {
+	var (
+		errs        []error
+		versionInfo Version
+	)
+
+	if !o.ServerOnly {
+		clientVersion := version.Get()
+		versionInfo.ClientVersion = &clientVersion
+	}
+
+	if !o.ClientOnly {
+		if o.discoveryClient != nil {
+			// Always request fresh data from the server
+			o.discoveryClient.Invalidate()
+			serverVersion, err := o.discoveryClient.ServerVersion()
+			errs = append(errs, err)
+			versionInfo.ServerVersion = serverVersion
+		}
+		if o.oClient != nil {
+			openshiftVersion, err := o.oClient.ClusterOperators().Get("openshift-apiserver", metav1.GetOptions{})
+			errs = append(errs, err)
+			for _, ver := range openshiftVersion.Status.Versions {
+				if ver.Name == "operator" {
+					// FIXME: openshift-apiserver DOES NOT report version
+					versionInfo.OpenShiftVersion = ver.Version
+				}
+			}
+		}
+	}
+
+	switch o.Output {
+	case "":
+		if o.All {
```

Comment:
is this the same logic as upstream to show git versions if all=false?

Reply:
upstream defaults to showing git versions, and `--short` to only show version number.  Here I'm defaulting to _not_ showing git versions and only showing number, unless you pass `--all` (could make it `--long` or could switch back to defaulting to show long form, but the OpenShift Server version _only_ has a short form, so it looks a bit weird.  Plus, I think it makes more sense to default to the short form.

Reply:
I'm guessing you'd prefer if I default to the long (what kubectl currently does) no matter if the short OpenShift server version doesn't match.  I'll hold out until you confirm, though.

Reply:
I've modified to keep same as kubectl - defaults to full information w/ flag for --short